### PR TITLE
Update Akka Analytics Cassandra to Apache Spark 1.6.1, Akka Persistence Cassandra to 0.14, Spark Cassandra Connector to 1.5.0

### PR DESCRIPTION
**Pull request by zapletal-martin**

### CODE REVIEW

File: akka-analytics-kafka/build.sbt [hunk 1]
```diff
@@ -5,12 +5,13 @@ fork in Test := true
 parallelExecution in Test := false
 
 libraryDependencies ++= Seq(
-  "com.github.krasserm" %% "akka-persistence-kafka" % "0.4",
-  "com.github.krasserm" %% "akka-persistence-kafka" % "0.4" % "test" classifier "tests",
-  "org.apache.curator"   % "curator-test"           % "2.7.1" % "test",
-  "org.apache.spark"    %% "spark-streaming"        % "1.4.0",
-  "org.apache.spark"    %% "spark-streaming-kafka"  % "1.4.0",
-  "org.apache.kafka"    %% "kafka"                  % "0.8.2.1" excludeAll(
+  "com.typesafe.akka"   %% "akka-persistence-experimental" % "2.3.4",
```
Comment (zapletal-martin) [inline akka-analytics-kafka/build.sbt:8]: There is a dependency Akka Analytics Kafka -> Akka Persistence Kafka -> Akka Testkit -> Akka Persistence (old version). Before updating Akka Analytics Kafka Akka Testkit and Akka Persistence Kafka would need updating too.

**Comment by krasserm** [comment]
Hi Martin, thanks a lot for the pull request, here's my +1 on all the changes. Some tests failed with a timeout though:

```
[error] Failed tests:
[error]     akka.analytics.cassandra.BatchProcessingSpec
[error]     akka.analytics.cassandra.StreamProcessingSpec
[error]     akka.analytics.cassandra.CustomSerializationSpec
```

Details:

```
[info] - must expose journaled events as RDD *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (10 seconds) during expectMsg while waiting for A-1
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:349)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:335)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:734)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply(ProcessingSpec.scala:70)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply(ProcessingSpec.scala:70)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:160)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(ProcessingSpec.scala:70)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(ProcessingSpec.scala:65)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(ProcessingSpec.scala:65)
[info]   ...

...

[info] - must expose journaled events as RDD *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (10 seconds) during expectMsg while waiting for A-1
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:349)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:335)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:734)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply(ProcessingSpec.scala:70)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply(ProcessingSpec.scala:70)
[info]   at scala.collection.immutable.Range.foreach(Range.scala:160)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(ProcessingSpec.scala:70)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(ProcessingSpec.scala:65)
[info]   at akka.analytics.cassandra.ProcessingSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(ProcessingSpec.scala:65)
[info]   ...

...

[info] - must support custom serialization *** FAILED ***
[info]   java.lang.AssertionError: assertion failed: timeout (10 seconds) during expectMsg while waiting for ExamplePayload(a)
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:349)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:335)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:734)
[info]   at akka.analytics.cassandra.CustomSerializationSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(CustomSerializationSpec.scala:98)
[info]   at akka.analytics.cassandra.CustomSerializationSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(CustomSerializationSpec.scala:94)
[info]   at akka.analytics.cassandra.CustomSerializationSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(CustomSerializationSpec.scala:94)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   ...
```

Can you reproduce?

**Comment by krasserm** [comment]
Do you plan to make further additions/changes to this PR? 

Regarding the akka-persistence-kafka dependency, I currently don't have time to work on it. I also found using Kafka for Akka Persistence and Eventuate problematic, as described in [this thread](http://search-hadoop.com/m/uyzND18ByiaBoj5v1), see also https://github.com/krasserm/akka-persistence-kafka/issues/20 and https://github.com/RBMHTechnology/eventuate/issues/59 for details.

Do you have any plans supporting Kafka as Akka Persistence storage backend? I would be interested in joining the efforts ...

**Comment by zapletal-martin** [comment]
Hi Martin,

thanks for your comments, I have only now noticed them. Let me read through your feedback and get back to you in in next couple of days with more details.

I have seen the same test failures and increasing the timeout helped on my machine. I will investigate further because it seems to keep occurring.

I do not have much insight into Akka Persistence Kafka. What is the current state and what do you think are the next steps in its evolution and future?

Thanks again,
Martin

**Comment by krasserm** [comment]
> I do not have much insight into Akka Persistence Kafka. What is the current state and what do you think are the next steps in its evolution and future?

It is still based on the old Java producer, switching to the new producer is problematic as explained in https://github.com/krasserm/akka-analytics/pull/5#issuecomment-222418548. The next step would be to get it working with the new producer but only if it can be extended/modified to support atomic batch writes to topic partitions. This is currently work in progress (see recent discussions in https://github.com/krasserm/akka-persistence-kafka/issues/20).

**Comment by zapletal-martin** [comment]
It is the setup phase (persisting) taking too long, not the actual read. I have increased the timeout, let's see if it helps. It is difficult to debug, because it consistently succeeds on my machine. Is there a CI for the project I am not seeing?

As for akka persistence kafka - it seems the discussion you referenced is moving in the same way as we discussed in the cassandra implementation. I am more than happy to help, although the topic seems to be quite active already :)

**Comment by krasserm** [comment]
> I have increased the timeout, let's see if it helps. It is difficult to debug, because it consistently succeeds on my machine

I didn't setup Travis CI for the project yet but the increased timeout is now working locally for me as well.

> As for akka persistence kafka - it seems the discussion you referenced is moving in the same way as we discussed in the cassandra implementation. I am more than happy to help ...

Thanks for offering help, I highly appreciate it as I'm also planning to support a Kafka storage backend for Eventuate. There will likely be many common problems to solve ... so I'm happy to actively support either of these initiatives 😃

**Comment by krasserm** [comment]
How do you want to proceed with this PR? Squash, merge and release or do you plan further additions?

**Comment by zapletal-martin** [comment]
I plan further additions in near future, e.g. updating the Kafka part (when Akka Persistence Kafka is updated). There are some other interesting options too, such as directly supporting DataFrame/DataSet apis or exposing a Cassandra stream of all events.

I think this PR is ready to go. Do you want me to squash? 

Releasing a new version may be a good idea, because the other aforementioned changes may take a while and this PR will make akka analytics work with the latest spark/akka persistence.

**Comment by krasserm** [comment]
> I plan further additions in near future, e.g. updating the Kafka part (when Akka Persistence Kafka is updated). 

Glad to hear that!

> There are some other interesting options too, such as directly supporting DataFrame/DataSet apis or exposing a Cassandra stream of all events.

I'm currently working on that in context of a Spark adapter (batch and stream) for Eventuate. I'm happy to contribute/discuss my findings should you be interested.

> I think this PR is ready to go. Do you want me to squash? 

Yes, please. I'll cut a release then.

**Comment by zapletal-martin** [comment]
Awesome, more than happy to help with that effort or port your work to akka analytics. I will reach out via email so we can discuss and avoid work duplication.

**Pull request merged**